process.go

```diff
@@ -0,0 +1,183 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"io"
+
+	"errors"
+)
+
+// == Direct input processing ========================================
+
+type renderResult struct {
+	err     error
+	inPath  string
+	outPath string
+}
+
+func processInputFiles(stringTemplate string, inFiles []string, outFiles []string, g *Gomplate) error {
+	inFiles, err := readInputs(stringTemplate, inFiles)
+	if err != nil {
+		return err
+	}
+
+	if len(outFiles) == 0 {
+		outFiles = []string{"-"}
+	}
+
+	results := make(chan *renderResult)
+	defer close(results)
+
+	for n, input := range inFiles {
+		go func(idx int, input string) {
+			err := renderTemplate(g, input, outFiles[idx])
+			results <- &renderResult{err, input, outFiles[idx]}
+		}(n, input)
+	}
+
+	return waitAndEvaluateResults(results, len(inFiles))
+}
+
+// == Recursive input dir processing ======================================
+
+func processInputDir(inputDir string, outDir string, g *Gomplate) error {
+	results := make(chan *renderResult)
+	defer close(results)
+	nr := processDir(inputDir, outDir, g, results, 0)
+	return waitAndEvaluateResults(results, nr)
+}
+
+func processDir(inPath string, outPath string, g *Gomplate, results chan *renderResult, nr int) int {
+	inPath = filepath.Clean(inPath)
+	outPath = filepath.Clean(outPath)
+
+	// assert tha input path exists
+	si, err := assertDir(inPath)
```

Comment:
Just a tip - the example in https://golang.org/pkg/io/ioutil/#ReadDir is editable and runnable to help verify these sorts of things.

In particular, this code:

```go
package main

import (
	"io/ioutil"
	"log"
)

func main() {
	_, err := ioutil.ReadDir("foo")
	if err != nil {
		log.Fatal(err)
	}
}
```

returns this error:
```
2009/11/10 23:00:00 open foo: No such file or directory
```

I think that's plenty clear enough 🙂